# EDA: Diagnosing Diabetes

### Goal of the Project
To inspect, clean, and validate the data using the EDA approach.

### Resource
This [dataset](https://www.kaggle.com/uciml/pima-indians-diabetes-database) is from the National Institute of Diabetes and Digestive and Kidney Diseases. It contains the following columns:

- `Pregnancies`: Number of times pregnant
- `Glucose`: Plasma glucose concentration per 2 hours in an oral glucose tolerance test
- `BloodPressure`: Diastolic blood pressure
- `SkinThickness`: Triceps skinfold thickness
- `Insulin`: 2-Hour serum insulin
- `BMI`: Body mass index
- `DiabetesPedigreeFunction`: Diabetes pedigree function
- `Age`: Age (years)
- `Outcome`: Class variable (0 or 1)

## Initial Inspection

In [16]:
import pandas as pd
import numpy as np

# Load the data
diabetes_data = pd.read_csv("diabetes.csv")
diabetes_data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age Outcome  
0                     0.627   50       1  
1                     0.351   31       0  
2                     0.672   32       1  
3                     0.167   21       0  
4                     2.288   33       1

In [17]:
# Get the number of columns
print(len(diabetes_data.columns))
# Get the number of rows
print(len(diabetes_data))

9
768


## Further Inspection

In [18]:
# Find whether columns contain null values
print(diabetes_data.isnull().sum())
# Perform summary statistics
diabetes_data.describe()


Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age  
count  768.000000                768.000000  768.000000  
mean    31.992578                  0.471876   33.240885  
std      7.884160                  0.331329   11.760232  
min      0.000000                  0.078000   21.000000  
25%     27.300000                  0.243750   24.000000  
50%     32.000000                  0.372500   29.000000  
75%     36.600000                  0.626250   41.000000  
max     67.100000                  2.420000   81.000000


In addition to the `0` values that show up for the columns above, there appear to be additional outliers, such as:

- The maximum value of the `Insulin` column is 846, which is abnormally high.
- The maximum value of the `Pregnancies` column is 17. While having 17 pregnancies is not impossible, this case might be something to look further into to determine its accuracy.


In [19]:
# Replace instances of 0 with NaN
diabetes_data[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']] = diabetes_data[[
    'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']].replace(0, np.NaN)

In [20]:
# Find whether columns contain null values after replacements are made
print(diabetes_data.isnull().sum())

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64


In [21]:
# Perform summary statistics
diabetes_data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  763.000000     733.000000     541.000000  394.000000   
mean      3.845052  121.686763      72.405184      29.153420  155.548223   
std       3.369578   30.535641      12.382158      10.476982  118.775855   
min       0.000000   44.000000      24.000000       7.000000   14.000000   
25%       1.000000   99.000000      64.000000      22.000000   76.250000   
50%       3.000000  117.000000      72.000000      29.000000  125.000000   
75%       6.000000  141.000000      80.000000      36.000000  190.000000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age  
count  757.000000                768.000000  768.000000  
mean    32.457464                  0.471876   33.240885  
std      6.924988                  0.331329   11.760232  
min     18.200000                  0.078000   21.000000  
25%     27.500000                  0.243750   24.000000  
50%     32.300000                  0.372500   29.000000  
75%     36.600000                  0.626250   41.000000  
max     67.100000                  2.420000   81.000000

We should take a look at the rows with missing data to get a better idea of the reason.

In [22]:
# Print rows with missing values
diabetes_data[diabetes_data.isnull().any(axis=1)]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6    148.0           72.0           35.0      NaN  33.6   
1              1     85.0           66.0           29.0      NaN  26.6   
2              8    183.0           64.0            NaN      NaN  23.3   
5              5    116.0           74.0            NaN      NaN  25.6   
7             10    115.0            NaN            NaN      NaN  35.3   
..           ...      ...            ...            ...      ...   ...   
761            9    170.0           74.0           31.0      NaN  44.0   
762            9     89.0           62.0            NaN      NaN  22.5   
764            2    122.0           70.0           27.0      NaN  36.8   
766            1    126.0           60.0            NaN      NaN  30.1   
767            1     93.0           70.0           31.0      NaN  30.4   

     DiabetesPedigreeFunction  Age Outcome  
0                       0.627   50       1  
1                       0.351   31       0  
2                       0.672   32       1  
5                       0.201   30       0  
7                       0.134   29       0  
..                        ...  ...     ...  
761                     0.403   43       1  
762                     0.142   33       0  
764                     0.340   27       0  
766                     0.349   47       1  
767                     0.315   23       0  

[376 rows x 9 columns]

Every single row with at least one missing value also has a missing value in the `Insulin` column. This is a clue as to why the data is missing. If patients did not have their insulin measured, they aldo didn't have other measurements taken.

In [23]:
# Print data types using .info() method
diabetes_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   763 non-null    float64
 2   BloodPressure             733 non-null    float64
 3   SkinThickness             541 non-null    float64
 4   Insulin                   394 non-null    float64
 5   BMI                       757 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    object 
dtypes: float64(6), int64(2), object(1)
memory usage: 54.1+ KB


To figure out why the `Outcome` column is of type `object` (string) instead of type `int64`, we should print out the unique values in the `Outcome` column.

In [24]:
# Check unique values of Outcome column
diabetes_data["Outcome"].unique()


array(['1', '0', 'O'], dtype=object)

To be able to manipulate data, we need to replace `O` with `0` looping through all the values in the column.

In [25]:
# Check the columns to loop through
diabetes_data.columns
# Loop through Outcome to replace the data values
for index in range(0, len(diabetes_data["Outcome"])):
    # Make each value a string
    string = str(diabetes_data["Outcome"].iat[index])
    # Remove a letter by replacing
    replace = string.replace("O", "0")
    # Perform replacement of each value in the column
    diabetes_data["Outcome"].iat[index] = replace


In [26]:
# Check unique values of Outcome column
diabetes_data["Outcome"].unique()

array(['1', '0'], dtype=object)

In [27]:
# Change data type into numerical
diabetes_data["Outcome"] = pd.to_numeric(diabetes_data["Outcome"])

To be able to perform more reliable manipulations, we should fill in the blank spacies with a `mean` function. 

In [28]:
# Get the mean value of Insulin column
insulin_mean = diabetes_data["Insulin"].mean(skipna=True)
# Replace NaN values with the new value
diabetes_data = diabetes_data.fillna(insulin_mean)
diabetes_data


Pregnancies  Glucose  BloodPressure  SkinThickness     Insulin   BMI  \
0              6    148.0           72.0      35.000000  155.548223  33.6   
1              1     85.0           66.0      29.000000  155.548223  26.6   
2              8    183.0           64.0     155.548223  155.548223  23.3   
3              1     89.0           66.0      23.000000   94.000000  28.1   
4              0    137.0           40.0      35.000000  168.000000  43.1   
..           ...      ...            ...            ...         ...   ...   
763           10    101.0           76.0      48.000000  180.000000  32.9   
764            2    122.0           70.0      27.000000  155.548223  36.8   
765            5    121.0           72.0      23.000000  112.000000  26.2   
766            1    126.0           60.0     155.548223  155.548223  30.1   
767            1     93.0           70.0      31.000000  155.548223  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]